### DISCUSSION TITLE
maybe add a tokenizer and token enumerator to core lib to make it easy to build recursive descendands parsers

### DISCUSSION THREAD

**Discussion by furesoft** [rzubek/CSLisp#25] [Ideas]

**Comment by rzubek** [comment]
Hmm, some examples would be appreciated!

**Comment by furesoft** [comment]
like an regex based tokenizer:

`(def builder (make-tokenizer-builder))`
`(add-lexeme builder "(" 'OpenParen)`
`(add-lexeme builder ")" 'CloseParen)`
`(add-lexeme builder "[0-9]+" 'Number)`

`(def tokenizer (make-tokenizer builder))`

`(def tokens (tokenize tokenizer "()"))`

**Comment by furesoft** [comment]
and  an enumerator to get the tokens:


`(peek-token tokens) -> gives next token, but dont consume it (lookahead)`

`(consume-token tokens) -> gives next token`

**Comment by furesoft** [comment]
a token is a pair (symbol kind, string value)

**Comment by furesoft** [comment]
what do you think?

**Comment by rzubek** [comment]
This sounds very appealing. Would you like to take a stab at implementing?

**Comment by furesoft** [comment]
it would make sense to store it in another package. we need a easier way to create packages on the clr side, maybe through attributes?

[LispPackage("core.vector")]
public static class VectorFunctions {

}

[LispPackage("core.parsing.tokenizer")]
public static class TokenizerFunctions {

}

**Comment by rzubek** [comment]
Yes, I think it makes sense to put the tokenizer in a separate package. Vector can probably stay in core, it's foundational enough!

**Comment by furesoft** [comment]
do you agree with the attribute to add functions to packages?

it would be more readable if we have many more functions and packages. I could implement the architecture and some sample packages like the tokenizer

**Comment by rzubek** [comment]
The attribute sounds great

**Comment by rzubek** [comment]
In this case, can I recommend a two step approach? First adding the attribute and refactoring how primitives get initialized. And then a separate task which is the work on the tokenizer itself.

**Comment by furesoft** [comment]
i will implement an architecture that collect all functions. all other stuff is implemented by the functions and the values are controlled by the runtime.

and for the tokenizer do you mean something like when a token is a boolean it should be automaticly converted to a boolean?

I could add something like initializer callbacks that conberts the string to a primitive and some basic implementations for it

**Comment by rzubek** [comment]
No, I meant generally what you suggested in the title of this issue ("maybe add a tokenizer and token enumerator"), not any specific kind of tokenizer.

Please feel free to add a tokenizer for your purposes, but let's keep that as a separate task, after the primitives refactor, so that primitives can have their own pull request.
